File: tests/testImageEncodeData.cpp
//
// Created by Andrei Costinescu (andreicostinescu96@gmail.com) on 29-May-21.
//

#include <comm/data/DataCollectionWithOpenCV.h>
#include <comm/data/ImageEncodeDataWithOpenCV.h>
#include <iostream>

using namespace comm;
using namespace std;

int main() {
    cout << "Hello World!" << endl;

    cv::Mat lena = cv::imread("../../data/Lena.png");
    DataCollectionWithOpenCV data;
    ImageEncodeDataWithOpenCV image(lena, 0, ImageEncodeData::JPEG), *decodedImage;
    decodedImage = dynamic_cast<ImageEncodeDataWithOpenCV *>(data.get(MessageType::IMAGE_ENCODE));
    cout << "Image size: " << lena.rows << "x" << lena.cols << ": " << lena.type()
         << "; raw image bytes = " << image.getImageBytesSize() << endl;

    Buffer b;
    bool serializeDone, deserializeDone;
    for (int encoding = 0; encoding <= 2; encoding++) {
        int state = 0;
        cout << "Expecting message type: " << messageTypeToString(decodedImage->getMessageType()) << endl;
        image.setEncoding(ImageEncodeData::Encoding(encoding));
        while (true) {
            serializeDone = image.serialize(&b, 0, false, true);
            deserializeDone = decodedImage->deserialize(&b, 0, false, true);
            cout << "Serialize state: " << state << ", buffer content size = " << b.getBufferContentSize() << endl;

            assert (deserializeDone == serializeDone);
            if (serializeDone) {
                break;
            }
            state++;
        }

        cout << "Encoded with " << ImageEncodeData::convertEncodingToString(image.getEncoding()) << " and decoded with "
             << ImageEncodeData::convertEncodingToString(decodedImage->getEncoding()) << endl;

        cv::imshow("Lena decoded", decodedImage->getImage());
        cv::waitKey(0);

        assert (decodedImage->getID() == image.getID());
        cout << "Different pixels in image per channel " << cv::sum(decodedImage->getImage() != image.getImage()) / 255
             << endl;
    }

    cout << "Test finished successfully!" << endl;

    return 0;
}

File: python/tests/testImageEncodeData.py
import cv2 as cv
import numpy as np
from comm.comm_data.DataCollection import DataCollection
from comm.comm_data.ImageEncodeData import ImageEncodeData
from comm.comm_data.MessageType import MessageType
from comm.comm_utils.Buffer import Buffer


def test():
    lena = cv.imread("/home/andrei/CLionProjects/Comm/data/Lena.png")
    print(lena.shape)
    image = ImageEncodeData(lena, 0, ImageEncodeData.Encoding.JPEG)
    data = DataCollection()
    decodedImage = data.get(MessageType.IMAGE_ENCODE)  # type: ImageEncodeData
    b = Buffer()

    for encoding in [ImageEncodeData.Encoding.JPEG, ImageEncodeData.Encoding.PNG, ImageEncodeData.Encoding.TIFF]:
        state = 0
        print("Expecting message type:", MessageType.messageTypeToString(decodedImage.getMessageType()))
        image.setEncoding(encoding)
        while True:
            serializeDone = image.serialize(b, 0, False, True)
            deserializeDone = decodedImage.deserialize(b, 0, False, False)
            print("Serialize state:", state, ", buffer content size =", b.getBufferContentSize())

            assert (deserializeDone == serializeDone);
            if serializeDone:
                break
            state = state + 1

        print("Encoded with", ImageEncodeData.Encoding.convertEncodingToString(image.getEncoding()), "and decoded with",
              ImageEncodeData.Encoding.convertEncodingToString(decodedImage.getEncoding()))

        cv.imshow("Lena decoded", decodedImage.getImage())
        cv.waitKey(0);

        assert (decodedImage.getID() == image.getID())
        print("Different pixels in image per channel", np.sum(np.sum(decodedImage.getImage() != image.getImage(), axis=0), axis=0))

    print("Test finished successfully!")


if __name__ == "__main__":
    test()
